### PR TITLE
mobile: Make sync mode configurable

### PR DESCRIPTION
**Pull request by ashishb**
Make sync mode configurable. Keep Light as the default sync mode to
maintain backward-compatibility.

The mode can be configured by calling `NodeConfig.setSyncMode()`

Fixes issue https://github.com/ethereum/go-ethereum/issues/17979

### CODE REVIEW

File: mobile/geth.go [hunk 1]
```diff
@@ -38,6 +38,13 @@ import (
 	whisper "github.com/ethereum/go-ethereum/whisper/whisperv6"
 )
 
+// I am intentionally duplicating these constants different from downloader.SyncMode integer values, to ensure the
+// backwards compatibility where the mobile node defaults to LightSync.
+const SyncModeUnset = 0 // will be treated as SyncModeLightSync
```
Comment (holiman) [inline mobile/geth.go]: In golang, it's common to use `iota` for these things: https://github.com/golang/go/wiki/Iota

**Comment by karalabe** [comment]
The code is still constructing a light client node: https://github.com/ethereum/go-ethereum/blob/master/mobile/geth.go#L164

**Comment by ligi** [comment]
I really do not see a use-case for something else than light sync on mobile. I would not do it just because it is possible. @ashishb what is your use-case?

File: mobile/geth.go (hunk 1 above)
Comment (ashishb) [reply mobile/geth.go]: Thanks. Fixed.

Comment (ashishb) [reply mobile/geth.go]: Thanks. Fixed.

**Comment by ashishb** [comment]
> Also, I see the addition of a new method, but I don't actually see how this is used anywhere?

My bad. This PR was incomplete, I was moving and removing the code from a fork and missed adding the actual usage. This has been fixed now.

> The code is still constructing a light client node: https://github.com/ethereum/go-ethereum/blob/master/mobile/geth.go#L164

Oversight on my part. Fixed, please take a look.

**Comment by ashishb** [comment]
> I really do not see a use-case for something else than light sync on mobile. I would not do it just because it is possible. @ashishb what is your use-case?

@ligi The use-case is less applicable to the mainnet since it is pretty big but for the private chains as well as the testnet where syncing does not take forever, this is useful.

**Comment by karalabe** [comment]
So, we've been debating about this PR, and we don't really see why anyone would want to run a full node on a mobile. Rinkeby is a testnet, but even that is 20 something GB. Goerli might still run now on a mobile, but it will definitely not after a few more months.

To get this PR in, you'd need to give us a goon enough reason where this can be useful.

**Comment by ashishb** [comment]
@karalabe The idea is to make this configurable, so that, the choice is made by the user without changing defaults. For example, we have forked Geth at Celo and added a new ultralight sync mode - https://github.com/celo-org/celo-blockchain. By making it configurable, it is possible for someone to add a new mobile-focused light sync mode to Geth.

**Comment by fjl** [comment]
As explained in the [corresponding issue](https://github.com/ethereum/go-ethereum/issues/17979#issuecomment-572484798), making the sync mode configurable such that full sync can be enabled on phones isn't something we want. There would be no issue with adding a separate flag for ultralight sync though. LES does support setting ultralight mode, so maybe that could be exposed in mobile instead.

**Pull request closed without merging**